File: ruby/tut5.rb
#!/usr/bin/env ruby

# XMMS2 - X Music Multiplexer System
# Copyright (C) 2003-2006 XMMS2 Team

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.

# This file is a part of the XMMS2 client tutorial #6.

# For the purposes of this tutorial, the client will print the ID of the
# currently playing song. Time to play your favorite song again.

# This is the async version of tutorial 2. You will need Ruby-GLib2 to run this
# tutorial. However, GLib is by no means the only event loop you can use for
# an asynchronous client. GLib is chosen since you need it to compile XMMS2
# anyway.
require 'xmmsclient'
# XMMS2 comes with a few mainloop integration tools we can use. Even better is
# that they are bound in Ruby. Here is the one for GLib2.
require 'xmmsclient_glib'
# It doesn't matter what order you require GLib. Even if it's required after
# xmmsclient_glib, the clientlib will still integrate properly with the GLib
# mainloop. To prove I'm not lying, the require statement is here.
require 'glib2'

begin
	xmms = Xmms::Client.new('tutorial6').connect(ENV['XMMS_PATH'])
rescue Xmms::Client::ClientError
	puts 'Failed to connect to XMMS2 daemon.'
	puts 'Please make sure xmms2d is running and using the correct IPC path.'
	exit
end

# GLib mainloop integration is a breeze with the tools provided for GLib in
# XMMS2. Just one function call sets it all up.
xmms.add_to_glib_mainloop

# Here we can initialize a GLib::MainLoop instance for later use. Ignore the
# arguments passed, they're unimportant boilerplate stuff. Also note that the
# mainloop is instantiated after the clientlib integrates itself with the
# mainloop. This is like above with the require statements.
ml = GLib::MainLoop.new(nil, false)

# So you've probably looked below and already noticed what's different here
# from sync clients. The 'wait' method isn't used, but a mysterious 'notifier'
# method stands in its place. The 'notifier' method takes a block which seems
# to contain code suspiciously similar to tut2. But that is where the
# similarities end. Unlike the sync examples, if you were to stop executing
# code after this block, nothing would be printed.

# This is where all that jabber above about mainloops and GLib fits in. Async
# code tells the server to wait in a different way. When a command is issued,
# it is added to a queue. The mainloop repeatedly calls a clientlib method to
# determine whether or not there are commands in this queue. If there are, then
# another clientlib method is called to send this command and tell the server
# we are awaiting a response. The server sends a response which gets put in
# another queue. Then another method is called that checks if any responses are
# waiting in the input queue. If there is, a value is extracted from the
# result sennt by the server. That value is then passed to the block.

# That may have been a complicated way of explaining things, so look at it this
# way. The mainloop does all the waiting on the server instead of the 'wait'
# method and when results come back, a callback is called. Ruby's blocks make
# this an absolute joy for us.
xmms.playback_current_id.notifier do |id|
	if(id == 0)
		puts 'There is no current ID. XMMS2 is probably not playing anything.'
	else
		puts "Currently playing ID: #{id}"
	end

	# The ID should only appear once. So we quit the mainloop here instead of
	# letting the program run indefinitely.
	ml.quit
end

puts 'You will notice that this message is printed before the ID.'

# The above block of text will be printed before the ID because the commands
# we've issued to the server have only been added to the queue. When the
# mainloop starts, they are actually sent to the server and waited on. So the
# callback is guaranteed to be called after this block of text is printed every
# time.
ml.run

# That's it! You know how to make an XMMS2 client in Ruby, so get to it!

File: c/tut5.c
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2006 XMMS2 Team
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  This file is a part of the XMMS2 client tutorial #5
 *  More about dicts and propdicts
 */

#include <stdlib.h>

/* include xmmsclient header */
#include <xmmsclient/xmmsclient.h>

/*
 * We will call this from xmmsv_dict_foreach later
 * in the program. Skip ahead to the main program and read
 * the information there first
 */
void
my_dict_foreach (const char *key, xmmsv_t *value, void *user_data)
{
	/*
	 * We get called for each entry in the dict.
	 * Here we need to decide how to print the value
	 * depending on its type and move on with life
	 */

	switch (xmmsv_get_type (value)) {
		case XMMSV_TYPE_NONE:
			/* nothing to do, empty value */
			break;
		case XMMSV_TYPE_INT32:
			{
				int val;
				xmmsv_get_int (value, &val);
				printf ("%s = %d\n", key, val);
				break;
			}
		case XMMSV_TYPE_STRING:
			{
				const char *val;
				xmmsv_get_string (value, &val);
				printf ("%s = %s\n", key, val);
				break;
			}
		default:
			break;
	}

}

/* We'll declare this one right below. */
void my_propdict_inner_foreach (const char *source, xmmsv_t *value,
                                void *user_data);

/*
 * This function is called for each pair of the key<->(source<->value)
 * propdict; we dispatch another foreach call to iterate over all the
 * source<->value pairs.
 */
void
my_propdict_foreach (const char *key, xmmsv_t *src_val_dict,
                     void *user_data)
{
	/* We pass the key along as user_data. */
	xmmsv_dict_foreach (src_val_dict, my_propdict_inner_foreach,
	                    (void *) key);
}

/*
 * This function is called for each inner pair source<->value of the
 * propdict. The parent key is received as user_data. We can now print
 * the full tuple from here!
 */
void
my_propdict_inner_foreach (const char *source, xmmsv_t *value,
                           void *user_data)
{
	/*
	 * We get called for each tuple in the propdict.
	 * Here we need to decide how to print the value
	 * depending on its type and move on with life
	 */
	const char *key = (const char *) user_data;

	switch (xmmsv_get_type (value)) {
		case XMMSV_TYPE_NONE:
			/* nothing to do, empty value */
			break;
		case XMMSV_TYPE_INT32:
			{
				int val;
				xmmsv_get_int (value, &val);
				printf ("%s:%s = %d\n", source, key, val);
				break;
			}
		case XMMSV_TYPE_STRING:
			{
				const char *val;
				xmmsv_get_string (value, &val);
				printf ("%s:%s = %s\n", source, key, val);
				break;
			}
		default:
			break;
	}

}

int
main (int argc, char **argv)
{
	/*
	 * The first parts of this program is
	 * commented in tut1.c
	 */
	xmmsc_connection_t *connection;
	xmmsc_result_t *result;
	xmmsv_t *return_value;

	/*
	 * Values that we need later
	 */

	connection = xmmsc_init ("tutorial1");
	if (!connection) {
		fprintf (stderr, "OOM!\n");
		exit (EXIT_FAILURE);
	}

	if (!xmmsc_connect (connection, getenv ("XMMS_PATH"))) {
		fprintf (stderr, "Connection failed: %s\n",
		         xmmsc_get_last_error (connection));

		exit (EXIT_FAILURE);
	}

	/*
	 * In tut3 we learned about dicts. But there is more to know on this
	 * topic. Some commands return more complex dict structures. I will
	 * try to explain them here.
	 *
	 * A normal dict contains key<->value mappings as normal, where
	 * the value is of a simple type (int, string). Getting
	 * values from this is straightforward (see tut3).
	 *
	 * On the other hand, xmmsc_medialib_get_info returns what we call a
	 * "property dicts", which is essentially of the form
	 * key<->(source<->value). In other words, each property key in the
	 * dict is mapped to a value which is also a dict. This inner dict
	 * maps property sources to property values.
	 * This allows the same property key to have multiple property
	 * values, one for each source. Like two "artists" or "titles".
	 * Property dicts (or propdicts) are primarily used by the medialib.
	 * In this case the source refers to the application which set
	 * the value of the property.
	 *
	 * Most of the time, you might not care and might just want one
	 * value per key, i.e. a simple key<->value mapping. The
	 * xmmsv_propdict_to_dict helper function converts a propdict into
	 * a simpler, flat dict (see the documentation to use it and use
	 * custom source preferences).
	 *
	 * In this tutorial, we will act as if we care about all sources
	 * and iterate over the propdict. Let me show you.
	 *
	 * First we retrieve the config values stored in the server and print
	 * them out. This is a normal dict.
	 */
	result = xmmsc_config_list_values (connection);
	xmmsc_result_wait (result);
	return_value = xmmsc_result_get_value (result);

	/*
	 * Iterating over a dict is done by calling a callback function for
	 * each key<->value pair in the dict.
	 */
	xmmsv_dict_foreach (return_value, my_dict_foreach, NULL);
	xmmsc_result_unref (result);

	/*
	 * Now get a "prop dict". Entry 1 should be the default clip
	 * we ship so it should be safe to request information
	 * about it.
	 */

	result = xmmsc_medialib_get_info (connection, 1);
	xmmsc_result_wait (result);
	return_value = xmmsc_result_get_value (result);

	/* now call xmmsc_result_prop_dict_foreach instead! */
	xmmsv_dict_foreach (return_value, my_propdict_foreach, NULL);
	xmmsc_result_unref (result);

	xmmsc_unref (connection);

	return (EXIT_SUCCESS);
}

